### ISSUE TITLE
Foxx legacy: import fails with indexes defined in repository.

### ISSUE THREAD

**Issue by bangert** [arangodb/arangodb#2119]
This is on 3.0.10 on Fedora 23.

Upon import of a legacy app with indexes defined in the repository, foxx-manager fails and the following shows up in the arango log.

```
2016-10-18T13:17:54Z [972] INFO Service at "/tsprovision" expects version "^2.8.0" and will run in legacy compatibility mode.
2016-10-18T13:17:54Z [972] ERROR Cannot compute Foxx service routes: ArangoError: failed to invoke module
2016-10-18T13:17:54Z [972] ERROR File: /tmp/arangod_OXTKzm/apps/tmp-972-1919842062/controllers/graphs.js
2016-10-18T13:17:54Z [972] ERROR Cause: TypeError: Cannot read property 'ensureIndex' of null
2016-10-18T13:17:54Z [972] ERROR     at child.<anonymous> (/usr/share/arangodb3/js/server/modules/@arangodb/foxx/legacy/repository.js:81:22)
2016-10-18T13:17:54Z [972] ERROR     at arrayEach (/usr/share/arangodb3/js/node/node_modules/lodash/lodash.js:451:11)
2016-10-18T13:17:54Z [972] ERROR     at Function.forEach (/usr/share/arangodb3/js/node/node_modules/lodash/lodash.js:8634:14)
2016-10-18T13:17:54Z [972] ERROR     at child.Repository (/usr/share/arangodb3/js/server/modules/@arangodb/foxx/legacy/repository.js:80:7)
2016-10-18T13:17:54Z [972] ERROR     at new child (/usr/share/arangodb3/js/node/node_modules/extendible/index.js:88:19)
2016-10-18T13:17:54Z [972] ERROR     at Object.<anonymous> (/tmp/arangod_OXTKzm/apps/tmp-972-1919842062/controllers/graphs.js:20:14)
2016-10-18T13:17:54Z [972] ERROR     at Module._compile (/usr/share/arangodb3/js/common/bootstrap/modules.js:593:6)
2016-10-18T13:17:54Z [972] ERROR     at Object.Module._extensions..js (/usr/share/arangodb3/js/common/bootstrap/modules.js:613:10)
2016-10-18T13:17:54Z [972] ERROR     at Module.load (/usr/share/arangodb3/js/common/bootstrap/modules.js:544:34)
2016-10-18T13:17:54Z [972] ERROR     at FoxxService.run (/usr/share/arangodb3/js/server/modules/@arangodb/foxx/service.js:397:12)
2016-10-18T13:17:54Z [972] ERROR     at Module.load (/usr/share/arangodb3/js/common/bootstrap/modules.js:547:21)
2016-10-18T13:17:54Z [972] ERROR     at FoxxService.run (/usr/share/arangodb3/js/server/modules/@arangodb/foxx/service.js:397:12)
2016-10-18T13:17:54Z [972] ERROR     at mountController (/usr/share/arangodb3/js/server/modules/@arangodb/foxx/legacy/routing.js:314:11)
2016-10-18T13:17:54Z [972] ERROR     at /usr/share/arangodb3/js/server/modules/@arangodb/foxx/legacy/routing.js:362:9
2016-10-18T13:17:54Z [972] ERROR     at Array.forEach (native)
2016-10-18T13:17:54Z [972] ERROR     at exports.routeService (/usr/share/arangodb3/js/server/modules/@arangodb/foxx/legacy/routing.js:361:36)
2016-10-18T13:17:54Z [972] ERROR     at exports.routeService (/usr/share/arangodb3/js/server/modules/@arangodb/foxx/routing.js:156:13)
2016-10-18T13:17:54Z [972] ERROR     at _validateService (/usr/share/arangodb3/js/server/modules/@arangodb/foxx/manager.js:1052:7)
2016-10-18T13:17:54Z [972] ERROR     at Object.replace (/usr/share/arangodb3/js/server/modules/@arangodb/foxx/manager.js:1306:3)
2016-10-18T13:17:54Z [972] ERROR     at Object.actions.defineHttp.callback.easyPostCallback.callback [as callback] (/usr/share/arangodb3/js/actions/_admin/foxx/app.js:127:26)
```

This is my repository file:

```
$ cat repositories/graphs.js 
'use strict';
var Foxx = require('org/arangodb/foxx');

module.exports = Foxx.Repository.extend({
  // Add your custom methods here
  indexes: [
    {
      type: 'hash',
      fields: ['name'],
      unique: true
    }
  ]
});
```

Any help greatly appreciated.
Is there a different way, i can define indexes in my app? Also, is it possible to have a mix of legacy and non-legacy code in the same app?

ArangoDB rocks - Thanks!
Thilo

**Comment by pluma** [comment]
This sounds like the `collection` property of the repository instance is missing.

This could happen if the first parameter to the `Repository` constructor is omitted or empty (e.g. when trying to use a collection that doesn't exist).

This doesn't seem like it's a problem with foxx-manager, the Foxx legacy mode or Foxx/ArangoDB itself.

To answer your other questions:

You can't have a mix of legacy and non-legacy code in the same app/service. You can use dependencies to export a JS API from your legacy app and access it from your non-legacy service (or vice versa) but legacy mode is all-or-nothing for each app/service.

You can create indexes using the `ensureIndex` method on collection objects in JS:

https://docs.arangodb.com/3.0/Manual/Indexing/WorkingWithIndexes.html#creating-an-index

This is actually what Repository did under the hood.

**Comment by bangert** [comment]
@pluma Thanks for your response. Here is how I instantiate the Graphs in the controller:

```javascript
var Graphs = require('../repositories/graphs');
var Graph = require('../models/graph');
var graphs = new Graphs(
  applicationContext.collection('graphs'),
  {model: Graph}
);
```
How should one go about doing this instead?

This works on 2.8, which is why I was mislead to blame it on legacy mode. ;-) 

Thank you for clearing up the question about mixing legacy and non-legacy mode code. 

kind regards
Thilo

**Comment by bangert** [comment]
```javascript
console.log(applicationContext.collection);
console.log(applicationContext.collection('graphs'));
console.log(applicationContext.collection('tsprovision_graphs'));
```

returns

```
2016-10-28T05:54:34Z [979] INFO /internal [Function: collection]
2016-10-28T05:54:34Z [979] INFO /internal null
2016-10-28T05:54:34Z [979] INFO /internal null
```

where `tsprovision_graphs` is the collection name in the db and `tsprovision` is the app name.

**Comment by pluma** [comment]
@bangert It looks like the collection wasn't created. `applicationContext.collection` doesn't create collections, it just looks them up by their prefixed name.

If the collection `tsprovision_graphs` does exist. Can you tell me what the mount path of the service is? The prefix is based on the mount path, not the name in the manifest.

**Comment by bangert** [comment]
```
foxx-manager replace . /tsprovision
```
is what i use - presumably giving the mount path of `/tsprovision`.

I think i may not have clearly stated earlier, that this is what i mean by 'import' above.

Given that the import succeeds if i take out the indexes attribute of the repository, and also the repository actually works, i am almost certainly back to blaming foxx-manager. it appears to fail before the app is installed, in a test environment of some sort...

**Comment by pluma** [comment]
@bangert does your service have a setup script? If so, could you add a `console.log` near the top of it and check whether that shows up in the logs?

Also, could you check whether the service you're replacing is running in development mode?

The problem is definitely that the collection doesn't exist at the time the constructor of the repository is invoked. That would normally indicate that the setup script doesn't create the collection or that the setup script is not being invoked.

In this case I'm not sure. My guess is that the service is being mounted before the setup script is executed. In this case, the message from your setup script should not show up in the logs when attempting to run `foxx-manager replace` on it.

**Comment by bangert** [comment]
running in development mode or not does not seem to make a difference. I do have setup script and it also run, but apparently only later.

when i disable the `indexes` attribute on the repository, the replace is finished successfully and i also see the console.log from setup.js.

when i enable `indexes` , the replace fails and i see nothing in the log from setup.js.

So an ordering issue? setup.js is run after app is mounted?

**Comment by pluma** [comment]
It seems that way. I'll have to look into this but I doubt it's specific to legacy mode.

**Comment by bangert** [comment]
well, i started porting the app to arangodb3. in the process i moved to a setup function that looks like this:

```
for (const localCollection of documentCollections) {
  const qualifiedName = module.context.collectionName(localCollection.name);
  var collection = db._collection(qualifiedName);
  if (!collection) {
    db._createDocumentCollection(qualifiedName);
    collection = db._collection(qualifiedName);
  } else if (module.context.isProduction) {
    console.warn(`collection ${qualifiedName} already exists. Leaving it untouched.`)
  }
  for (const index of localCollection.indexes) {
    collection.ensureIndex(index);
  }
}
```

which seems to work. so that solved my problem somewhat. If i can get the app to actually work now, then I am happy, as I would not have any legacy apps left :-P

**Comment by fceller** [comment]
@mpv1989 can we close this issue? Can you check with @pluma?

**Comment by pluma** [comment]
Further investigation has shown that this issue is caused by the "validation" logic which attempts to execute a service to detect any mount-time errors present in the code. The problem with this logic is that it also fails if the service relies on the setup script having already executed -- which should be normal supported behaviour.

As part of the cluster resilience effort we're replacing the mock execution "validation" step with a less error-prone version that merely parses the files without executing them. Additionally in legacy services the controller and export files will no longer be validated.

These changes are expected to be released as part of ArangoDB 3.2 but the feature branch has not been merged yet.

**Comment by pluma** [comment]
I'm closing this in favour of #2408, which covers the same broken validation step causing the observed behaviour.
